README.md

```diff
@@ -1,6 +1,1864 @@
+
+### ✅ Registration Rules
+
+When a user registers, we enforce **two key checks**:
+
+1. Address hasn’t registered before
+2. Commitment is unique (no duplicates)
+
+👉 If either fails → revert.
+👉 If another address tries to reuse the same commitment → reject as well.
+
+> 💡 Don’t forget to implement the necessary state variables.
+
+### 🌱 Inserting Into the Merkle Tree
+
+Once checks pass:
+
+- Insert the new leaf into the incremental Merkle tree (`LeanIMT`) using the commitment as leaf value (library is already imported in `Voting.sol`). Call `insert()` directly on it.
+- Store the Merkle tree in a state variable: `s_tree`.
+- Emit an event with the **leaf index** + **value**:
+  - **Index** = leaf’s position when inserted
+  - **Value** = the commitment
+
+> 💡 Explore the LeanIMT library itself to see exactly what happens inside `insert()`.
+
+<details>
+<summary><b>🧠 🦉 Guiding Questions</b></summary>
+
+<details>
+<summary>❓ Question 1</summary>
+
+What conditions must be true for someone to register?
+_(Hint: allowlist + not registered before)_
+
+</details>
+
+<details>
+<summary>❓ Question 2</summary>
+
+How will you detect and reject a reused commitment, even from another address?
+
+</details>
+
+<details>
+<summary>❓ Question 3</summary>
+
+Which **state variables** do you need to track both users and commitments?
+
+</details>
+
+<details>
+<summary>❓ Question 4</summary>
+
+When emitting `NewLeaf`, how will you determine the correct index?
+_(💡 Try `tree.size()`, but adjust carefully)_
+
+</details>
+
+After thinking through the guiding questions, have a look at the solution code!
+
+<details>
+<summary>👩🏽‍🏫 Solution Code</summary>
+
+```javascript
+///////////////////////
+/// State Variables ///
+///////////////////////
+
+/// Checkpoint 2 //////
+mapping(address => bool) private s_hasRegistered;
+mapping(uint256 => bool) private s_commitments;
+
+LeanIMTData private s_tree;
+
+//////////////////
+/// Functions ///
+//////////////////
+
+function register(uint256 _commitment) public {
+/// Checkpoint 2 //////
+if (!s_voters[msg.sender] || s_hasRegistered[msg.sender]) {
+revert Voting__NotAllowedToVote();
+}
+if (s_commitments[_commitment]) {
+revert Voting__CommitmentAlreadyAdded(_commitment);
+}
+s_commitments[_commitment] = true;
+s_hasRegistered[msg.sender] = true;
+s_tree.insert(_commitment);
+emit NewLeaf(s_tree.size - 1, _commitment);
+}
+
+```
+
+</details>
+</details>
+
+### 🔧 Before Testing
+
+Scroll down to the functions **`getVotingData()`** and **`getVoterData(address _voter)`** in your contract.
+
+👉 Uncomment everything below: `/// Checkpoint 2 ///`
+
+Then run:
+
+```javascript
+yarn test --grep "Checkpoint2"
+```
+
+### 🚀 Tests Passed? You’re Almost There!
+
+Great job! If your tests are passing, you’re just one step away from deployment! 🚀
+
+Before deploying, make one important change:
+
+1. Open **`00_deploy_your_voting_contract.ts`**
+2. Set your address as the `ownerAddress`
+3. Comment out `leanIMTAddress`
```

Comment:
Change step 3 to something like "Set LeanIMT library address (leanIMT.address) at line 62"?